src/main/java/fi/aalto/cs/apluscourses/dal/APlusExerciseDataSource.java

```diff
@@ -21,46 +21,41 @@
 import java.util.List;
 import java.util.Map;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 import org.json.JSONArray;
 import org.json.JSONObject;
 import org.json.JSONTokener;
 
-public class APlusExerciseDataSource extends ExerciseDataSource {
+public class APlusExerciseDataSource implements ExerciseDataSource {
 
   @NotNull
   private final Client client;
 
   @NotNull
-  private String apiUrl;
+  private final String apiUrl;
 
   @NotNull
   private final Parser parser;
 
   /**
    * Default constructor.
-   *
-   * @param authProvider Used for creating an authentication for this data source.
    */
-  public APlusExerciseDataSource(@NotNull AuthProvider authProvider) {
```

Comment:
ExerciseDataSource no longer gets Authentication in constructor but it's given separately for each call